gateway/src/node/client_handling/websocket/connection_handler/eth_events.rs

```diff
@@ -0,0 +1,210 @@
+        let mnemonic =
+            Mnemonic::from_str(&cosmos_mnemonic).expect("Invalid Cosmos mnemonic provided");
+        let wallet = DirectSecp256k1HdWallet::from_mnemonic(mnemonic)
+            .expect("Could not creat wallet from mnemonic");
+        let client_address = wallet
+            .try_derive_accounts()
+            .expect("Could not get account from wallet")
+            .into_iter()
+            .map(|account| account.address().clone())
+            .collect::<Vec<AccountId>>()
+            .get(0)
+            .expect("No account for wallet")
+            .clone();
+        let nymd_client = SigningNymdClient::connect_with_signer(nymd_url.as_str(), wallet)
+            .expect("Could not create nymd client");
+
+        ERC20Bridge {
+            contract: eth_contract(web3.clone()),
+            web3,
+            nymd_client,
+            client_address,
+        }
+    }
+
+    pub(crate) async fn verify_eth_events(
+        &self,
+        verification_key: PublicKey,
+    ) -> Result<(), RequestHandlingError> {
+        // It's safe to unwrap here, as we are guarded by a unit test that checks the event
+        // name constant against the contract abi
+        let event = self.contract.abi().event(ETH_EVENT_NAME).unwrap();
+        let filter = FilterBuilder::default()
+            .address(vec![self.contract.address()])
+            .topics(
+                Some(vec![event.signature()]),
+                Some(vec![H256::from(verification_key.to_bytes())]),
+                None,
+                None,
+            )
+            .from_block(BlockNumber::Earliest)
+            .to_block(BlockNumber::Latest)
+            .build();
+        // Get only the first event that checks out. If the client burns more tokens with the
+        // same verification key, those token would be lost
+        for l in self.web3.eth().logs(filter).await? {
+            let log = event.parse_log(web3::ethabi::RawLog {
+                topics: l.topics,
+                data: l.data.0,
+            })?;
+            let burned_event =
+                Burned::from_tokens(log.params.into_iter().map(|x| x.value).collect::<Vec<_>>())?;
+            if burned_event.verify(verification_key) {
+                return Ok(());
+            }
+        }
+
+        Err(RequestHandlingError::InvalidBandwidthCredential)
+    }
+
+    pub(crate) async fn verify_double_spending(
+        &self,
+        credential: &TokenCredential,
+    ) -> Result<(), RequestHandlingError> {
+        // It's ok to unwrap here, as the cosmos contract and denom are set correctly
+        let contract_address = AccountId::from_str(COSMOS_CONTRACT_ADDRESS).unwrap();
+        let coin = Coin {
+            denom: Denom::from_str(DENOM).unwrap(),
+            amount: Decimal::from(100000u64),
+        };
+        let fee = Fee::from_amount_and_gas(coin.clone(), Gas::from(500000));
+        let req = ExecuteMsg::LinkPayment {
```

Comment:
Perhaps to make it slightly nicer, you could move transaction generation (and fee estimation) to nymd client?

Reply:
That `NymdClient` looks more like a specific mixnet client, which is the reason I didn't try using it in the first place. But given that it simplifies the whole setup, I don't see any harm in re-using the `execute` part of it for this contract as well. However, I would prefer to avoid merging both contract APIs into it, especially since this code is meant to go away at some point.